### PR TITLE
Feature/trunk/refinery http

### PR DESCRIPTION
**Pull request by chfsx**
This is the followup PR for #2344, now based on trunk

### CODE REVIEW

**Comment by chfsx** [comment]
Copied from the old PR:

Please answer the following questions:
- [x] ilPropertyFormGUI: Can we really remove [this](https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-411e64d40db5eeb3d6a82486fdc5e607L503) chunk of code? Why don't we need it anymore?
- [x] ReflectionClass: What's the purpose of [`HTTP\Services::getWithMultipleArguments`](https://github.com/ILIAS-eLearning/ILIAS/pull/2344/files#diff-e50f175337eeff55968c5b4b2650f8c8R153)? Using the `ReflectionClass` is (deemed to be) a performance impact. Why do we take it here? We know the classes that we want to instantiate.
- [x] DirectArrayAccessDropInReplacement: What is the purpose or added benefit of `DirectArrayAccessDropInReplacement`? Using the `Refinery` internally won't do anything to make this more secure, at least regarding the current implementation. Transforming the contents to string basically is a non-operation from the perspective of the result, the contents in the request super globals are strings anyway.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] DI: Please do not remove the dependency injection of `GlobalHttpState`. This will make testing and reuse of the class harder. What's the benefit of creating dependencies inline?

Please implement the following changes:
- [x] Wrapper/Raw: In current proposal you suggest to add a `wrapper` method to the `GlobalHttpState`, which will give access to the methods that use the `Refinery` for data retrieval. These should become the new normal for almost all of the ILIAS code. The other methods, like `request`, on the `GlobalHttpState` on the other hand provide access to the raw request objects, which mostly should not be to used anymore. Thus we suggest to change the `GlobalHttpState` so that `request` and `cookieJar` return the objects that use the `Refinery` for access and only make the raw objects accessible behind `raw`. Maybe we could even give a name that hints at the unsafeness even more, like  `getUnsafeRawObjects`. An update of the existing usage should be straight forward.
- [x] HTTPService: With the changes here, `HTTPServices` seems to have become an odd name. Please rename to `RawHTTPObjects`.
- [x] RequestWrapper: This should be pretty generic for all kinds of inputs. Rename it to something like `KeyValueInput` and move it to the top-level of the Refinery. The same goes for `ArrayBasedRequestWrapper`, which could be renamed to `ArrayInput` and also be moved to the `Refinery`.
- [ ] Dicto: Add some Dicto rules that flag usages of the request super globals and `getUnsafeRawObject`.
- [ ] Tests: Please add automated tests that at least cover the core features of the HTTP library.

**Comment by chfsx** [comment]
also copied from the old PR: 

Thanks for the review and no problem that it takes a little longer (it took me a long time now, too :-) )

with pleasure I will go into your points:

>  ilPropertyFormGUI: Can we really remove this chunk of code? Why don't we need it anymore?

this was to check if post_max_size exceeded, but this leads to an error anyway.

> ReflectionClass: What's the purpose of HTTP\Services::getWithMultipleArguments? Using the ReflectionClass is (deemed to be) a performance impact. Why do we take it here? We know the classes that we want to instantiate.

This was to handle the internal instances a bit easier, but I can remove this as well

> DirectArrayAccessDropInReplacement: What is the purpose or added benefit of DirectArrayAccessDropInReplacement? Using the Refinery internally won't do anything to make this more secure, at least regarding the current implementation. Transforming the contents to string basically is a non-operation from the perspective of the result, the contents in the request super globals are strings anyway.

Its correct that this doesn't bring a direct improvement of security. But it's a way to handle code which accesses $_GET directly in the future. With this implementation all accesses to $_GET go through this implementation, it will be easy to prevent from doing this in the future (which should be done IMO). I highly recommend to use this.

> DI: Please do not remove the dependency injection of GlobalHttpState. This will make testing and reuse of the class harder. What's the benefit of creating dependencies inline?

**Comment by chfsx** [comment]
and now for the new PR:

> Wrapper/Raw: In current proposal you suggest to add a wrapper method to the GlobalHttpState, which will give access to the methods that use the Refinery for data retrieval. These should become the new normal for almost all of the ILIAS code. The other methods, like request, on the GlobalHttpState on the other hand provide access to the raw request objects, which mostly should not be to used anymore. Thus we suggest to change the GlobalHttpState so that request and cookieJar return the objects that use the Refinery for access and only make the raw objects accessible behind raw. Maybe we could even give a name that hints at the unsafeness even more, like getUnsafeRawObjects. An update of the existing usage should be straight forward.

I agree that this should be the new normal. however, we probably have to work with deprecated (and perhaps dicto), otherwise many places in ILIAS would already have to be adapted. of course, this could also be done, but it is difficult, because plugins, for example, could already rely on it.

> HTTPService: With the changes here, HTTPServices seems to have become an odd name. Please rename to RawHTTPObjects.

absolutely, I'll be happy to adjust that.

> RequestWrapper: This should be pretty generic for all kinds of inputs. Rename it to something like KeyValueInput and move it to the top-level of the Refinery. The same goes for ArrayBasedRequestWrapper, which could be renamed to ArrayInput and also be moved to the Refinery.

I'm not adapting that either. Unfortunately, I lost some of the groundwork by re-implementing the PR.

> Dicto: Add some Dicto rules that flag usages of the request super globals and getUnsafeRawObject.

at the moment i don't have a working dicto environment to test this, but i am still working on it.

> Tests: Please add automated tests that at least cover the core features of the HTTP library.

I would then supply the tests as a final adjustment.

File: src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+     * @inheritDoc
+     */
+    public function offsetGet($offset)
+    {
+        if (!$this->offsetExists($offset)) {
+            return null;
+        }
+
+        if (is_array($this->raw_values[$offset])) {
+            return $this->factory->to()->dictOf($this->factory->kindlyTo()->string())->transform($this->raw_values[$offset]);
```
Comment (mjansenDatabay) [inline src/HTTP/Wrapper/DirectArrayAccessDropInReplacement.php]: @chfsx Maybe I am wrong, but IMO this does not work with multi level POST array.

**Comment by chfsx** [comment]
@mjansenDatabay II implemented the topic we discussed:
- deprecation of "raw" methods
- renaming of Services
- KeyValueAccess in refinery AND injection of the transformation to use. please note that I extend that for a new SuperGlobalDropInReplacement to have the possibility to throw a exception if data is modified
- removal of statics and reflection

currently missing are tests and dicto-rules which I will deliver after your feedback to the current state. thanks a lot!

**Comment by chfsx** [comment]
@mjansenDatabay I implemented some test for the new part of the HTTP Service but I just fail to implement (or run) dictor-tests, not because they're broken but I just don't get a running environment for this.

**Comment by klees** [comment]
Hi @chfsx,

could you just post a ticket that demands the dicto rule and hand it to me?

Thanks!

**Comment by chfsx** [comment]
OC, see in https://mantis.ilias.de/view.php?id=31013

**Comment by kergomard** [comment]
Thank you for the work on the refinery! We discussed this PR in the Technical Board today and decided to merge without waiting for the dicto rules to not slow things down additionally.

**Pull request merged**